symbolic-il2cpp/src/usymlite.rs

```diff
@@ -0,0 +1,180 @@
+#[repr(C)]
+struct UsymLiteLine {
+    address: u64,
+    /// Reference to the managed source file name in the string table.
+    filename: u32,
+    /// Managed line number.
+    line: u32,
+}
+
+pub struct UsymLiteSymbols<'a> {
+    header: &'a UsymLiteHeader,
+    lines: &'a [UsymLiteLine],
+    /// The string table.
+    ///
+    /// This is a large slice of bytes with null-terminated C strings.
+    string_table: &'a [u8],
+}
+
+impl<'a> UsymLiteSymbols<'a> {
+    const MAGIC: &'static [u8] = b"sym-";
+
+    pub fn parse(buf: &'a [u8]) -> Result<UsymLiteSymbols<'a>> {
+        if buf.as_ptr().align_offset(8) != 0 {
+            // Alignment only really matters for performance really.
+            return Err(Error::msg("Data buffer not aligned to 8 bytes"));
+        }
+        if buf.len() < mem::size_of::<UsymLiteHeader>() {
+            return Err(Error::msg("Data smaller than UsymLiteHeader"));
+        }
+        if buf.get(..4) != Some(Self::MAGIC) {
+            return Err(Error::msg("Wrong magic number"));
+        }
+
+        // SAFETY: We checked the buffer is large enough above.
+        let header = unsafe { &*(buf.as_ptr() as *const UsymLiteHeader) };
+        if header.version != 2 {
+            return Err(Error::msg("Unknown version"));
+        }
+        let line_count: usize = header.line_count.try_into()?;
+        let stringtable_offset =
+            mem::size_of::<UsymLiteHeader>() + line_count * mem::size_of::<UsymLiteLine>();
+        if buf.len() < stringtable_offset {
+            return Err(Error::msg("Data smaller than number of line records"));
+        }
+
+        // SAFETY: We checked the buffer is at least the size_of::<UsymLiteHeader>() above.
+        let lines_ptr = unsafe { buf.as_ptr().add(mem::size_of::<UsymLiteHeader>()) };
+
+        // SAFETY: We checked the buffer has enough space for all the line records above.
+        let lines = unsafe {
+            let lines_ptr: *const UsymLiteLine = lines_ptr.cast();
+            let lines_ptr = ptr::slice_from_raw_parts(lines_ptr, line_count);
+            lines_ptr
+                .as_ref()
+                .ok_or(Error::msg("lines_ptr was null pointer!"))
+        }?;
+
+        let stringtable = buf
+            .get(stringtable_offset..)
+            .ok_or(Error::msg("No string table found"))?;
+        if stringtable.last() != Some(&0u8) {
+            return Err(Error::msg("String table does not end in NULL byte"));
+        }
+
+        Ok(Self {
+            header,
+            lines,
+            string_table: stringtable,
+        })
+    }
+
+    /// Returns a string from the string table at given offset.
+    ///
+    /// Offsets are as provided by some [`UsymLiteHeader`] and [`UsymLiteLine`] fields.
+    fn get_string(&self, offset: u32) -> Option<&'a CStr> {
+        // Panic if size_of::<usize>() < size_of::<u32>().
+        let offset: usize = offset.try_into().unwrap();
+        if offset >= self.string_table.len() {
```

Comment:
nit: i'm sure that inline type hints mitigate this problem, but `string_table` here feels like a real misnomer especially when used in this context because this comparison looks wrong at a glance: it looks as if you're checking the total number of entries in the string table against the offset, instead of the length of the table's raw contents.